tests/test-ptrace.c

```diff
@@ -178,7 +178,7 @@ main (int argc, char **argv)
 
   if (argc == 1)
     {
-      static char *args[] = { "self", "/bin/ls", "/usr", NULL };
+      static char *args[] = { "self", "/usr/bin/env", "ls", "/usr", NULL };
```

Comment:
I get why NixOS does things it does but can it really not put coreutils binaries in their standard location? This is just unnecessarily hostile, nobody needs multiple versions of ls to coexist...

Reply:
Well, that's one part of what you may call the stubbornness of the Nix philosophy. Nothing except for `/usr/bin/env` and `/bin/sh` (mandated by POSIX) is in their standard locations.

Reply:
Well, even `sh` isn't guaranteed to be in `/bin`:

> Applications should note that the standard PATH to the shell cannot be assumed to be either /bin/sh or /usr/bin/sh, and should be determined by interrogation of the PATH returned by getconf PATH, ensuring that the returned pathname is an absolute pathname and not a shell built-in.

-- https://pubs.opengroup.org/onlinepubs/9699919799/utilities/sh.html

Nor is `/usr/bin/env` mandated by POSIX, it's just a de-facto convention, same as `/bin/ls`, just NixOS decides to follow the former and not the latter.

Point is, NixOS already makes some compromises on its philosophy for practical reasons. I get why you would want to version some things, but not having things like `/bin/ls` in their standard location serves no practical benefit other than to break code that works on every other Unix and annoy developers with having to support this unnecessary arbitrary choice.

Reply:
The thing is, having exactly 2 exceptions is better than having an indeterminate number of them :stuck_out_tongue: Also the behavior of `/bin/ls` can and does change, depending on which coreutils version/implementation you are using. A great example can be found from recent effort to support uutils (alternative coreutils implementation written in Rust) in NixOS/nixpkgs#116274. Such a change would be virtually impossible to validate in other distributions.